File: leetcode/67. Add Binary/3.js
/**
 * Complexity (n = input string 1 length, k = input string 2 length)
 * Time complexity: O(max(n,k))
 * Space complexity: O(max(n,k))
 */
var addBinary = function ( a, b ) {
    let carry = 0;
    let result = [];
    for ( let i = 0; i < Math.max( a.length, b.length ); i++ ) {
        if ( i < a.length ) {
            carry += parseInt( a[a.length - i - 1] );
        }
        if ( i < b.length ) {
            carry += parseInt( b[b.length - i - 1] );
        }

        result.push( carry & 1 ? "1" : "0" );
        carry = carry & ( 1 << 1 ) ? 1 : 0;
    }

    if ( carry === 1 ) {
        result.push( "1" );
    }

    return result.reverse().join( '' );
};

File: leetcode/3. Longest Substring Without Repeating Characters/3.py
# Complexity (n = input length)
# Time complexity: O(n)
# Space complexity: O(ALPHABET_SIZE)

class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        window = set()
        max_len = 0
        slow_index = 0
        for fast_index, char in enumerate(s):
            while char in window:
                window.remove(s[slow_index])
                slow_index += 1
            window.add(char)

            max_len = max(max_len, len(window))

        return max_len
